roadmap/state-sync-protocol.md

```diff
@@ -0,0 +1,198 @@
+
+Such a protocol should require a minimal amount of additional work from Execution nodes, and in particular it should not involve them doing a fan-out of the state they generate. Instead, Execution nodes should only need to answer a bounded number of initial state requests, after which the further exchange of this data should be delegated among Access nodes themselves. This allows us to leverage the scability of the Access node layer and gain increased data dissemination robustness and decentralization.
+
+## Design
+
+### Data Transfer
+
+The core data exchange protocol we build on top of is called [Bitswap](https://docs.ipfs.io/concepts/bitswap/). At a high level, Bitswap's main responsibilities are to acquire chunks of data requested by the client from the network, and send chunks of data in its possession to other peers who want them.
+
+Bitswap functions with a few fundamental concepts:
+* A *block* represents a raw chunk of data.
+* Each block is uniquely identified by a [*content identifier (CID)*](https://docs.ipfs.io/concepts/content-addressing/), which is based on a cryptographic hash of the block's contents.
+* A *blockstore* is a data store for storing and retrieving blocks.
+* A *session* orchestrates the flow of messages with other peers for discovering and retrieving the individual blocks which comprise a single high-level object.
+
+The full details of the Bitswap protocol can be found [here](https://research.protocol.ai/publications/accelerating-content-routing-with-bitswap-a-multi-path-file-transfer-protocol-in-ipfs-and-filecoin/delarocha2021.pdf).
+
+> To avoid conflicting terminology, we will henceforth refer to blocks in Bitswap as *blobs*.
+
+### State Computation
+
+Execution nodes, upon executing the transactions in each block, will compute the following data:
+
+```protobuf
+message ExecutionData {
+    repeated Collection collections = 1;
+    repeated TransactionResult txResults = 2;
+    repeated Event events = 3;
+    repeated TrieUpdate trieUpdates = 4;
+}
+```
+
+They will then serialize each of these fields separately<sup>[1](#footnotes)</sup> and compute a root CID from the serialized data for each field as follows:
+
+1. Split the data into blobs of up to 1MB (the message size limit of the [underlying networking stack](https://docs.ipfs.io/concepts/libp2p/)).
+2. Compute the CID of each blob and store it in the blobstore.
+3. If the number of CIDs computed in the previous step is 1, then return this CID. Otherwise, concatenate the CIDs and repeat from step 1<sup>[2](#footnotes)</sup>.
+
+**[INSERT DIAGRAM HERE]**
+
+At the end of this process, we are left with four root CIDs (one for each field in the Execution Data). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.
```

Comment:
`Execution Results` _must_ be deterministic. If two ENs have the same result, they must generate the same `Execution Result`. This implies we need protocol definitions how elements `repeated TrieUpdate trieUpdates` are ordered.

Reply:
I think that *if* we end up deciding on having AN's be staked, and therefore the state sync protocol becomes part of the actual core protocol, then it's also possible to have Access Nodes do the verification, as described below in the Verification section.

On the other hand, if not, then yes verifier nodes will need to do the checking, and I have written a little bit on that in the Verification section as well. 

The high level idea is that instead of generating a single `ExecutionData` per block and creating a root CID(s) for that single `ExecutionData`, the Execution nodes would instead create a separate `ExecutionData` for each chunk. Therefore, instead of just one set of root CID's in the ExecutionResult, there would be `numberOfChunks` sets of root CID's. 

Then, when a verification node is verifying each chunk, they would recompute the `ExecutionData` for that chunk, and compare the root CID's they get vs the ExecutionResult's root CID's.

@AlexHentschel does this make sense? I will add this to the doc, if there are any suggestions on how to phrase this to make it more clear please let me know.

Reply:
> This implies we need protocol definitions how elements repeated TrieUpdate trieUpdates are ordered.

Yes, thanks for that flag. My assumption was that there is already a defacto ordering (namely, the order that the updates were applied when executing the block), but maybe this is incorrect. Can you clarify here?

Once this is clear, I will add more detail to the doc.

Reply:
> The high level idea is that instead of generating a single `ExecutionData` per block and creating a root CID(s) for that single `ExecutionData`, the Execution nodes would instead create a separate ExecutionData for each chunk.

👍  makes sense

>  defacto ordering (namely, the order that the updates occurred when executing the block)

The convention is that read and writes are deterministically ordered by their respective register ID. The ordering during execution is _part_ of the entropy that goes into creating the SPoCK proof. Therefore, we don't want to expose it as part of the result.